src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -937,6 +1073,104 @@ size_t runtime::gpu::CUDAEmitter::build_replace_slice(const GPURuntimeContext* c
+                                                  const std::array<std::string, 2>& dtypes,
+                                                  GPUShape result_shape,
+                                                  const std::set<size_t>& reduce_axes)
+{
+    // assumes NC{d1,...,dn} format
+    std::string kernel_name = "broadcast_" + join(dtypes, "_");
+    std::replace(kernel_name.begin(), kernel_name.end(), ' ', '_');
+
+    std::stringstream ss;
+    ss << kernel_name << "_s" << join(result_shape, "_") << "_r" << join(reduce_axes, "_");
+    auto hash = ss.str();
+
+    // check if the requested kernel is already an inserted primitive
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    // if the kernel has not been compiled, build it
+    auto compiled_kernel = ctx->compiled_kernel_pool->get(kernel_name);
+    if (compiled_kernel == nullptr)
+    {
+        codegen::CodeWriter writer;
+        writer << include_helpers();
+        runtime::gpu::CudaKernelBuilder::get_broadcast_op(
+            writer, kernel_name, dtypes, result_shape.size());
+        compiled_kernel = ctx->compiled_kernel_pool->set(kernel_name, writer.get_code());
+    }
+
+    // calculate strides
+    GPUShape strides = row_major_strides(result_shape);
+    // precacluate invariants for integer division via multiplication
+    std::vector<int> stride_magic;
+    std::vector<int> stride_shift;
+    for (int i = 0; i < strides.size(); i++)
+    {
+        int magic;
+        int shift;
+        std::tie(magic, shift) = idiv_magic_u64(strides[i]);
+        stride_magic.push_back(magic);
+        stride_shift.push_back(shift);
+    }
+    // calculate reduced tensor strides with 0s inserted for reduced axes
+    GPUShape reduced_shape = result_shape;
+    for (auto const& axis : reduce_axes)
+    {
+        reduced_shape[axis] = 1;
+    }
+    GPUShape reduced_strides = row_major_strides(reduced_shape);
+    for (auto const& axis : reduce_axes)
+    {
+        reduced_strides[axis] = 0;
+    }
+
+    GPUAllocator allocator = this->m_primitive_emitter->get_memory_allocator();
+    size_t idx_strides = allocator.reserve_argspace(strides.data(), strides.size() * sizeof(int));
+    size_t idx_stride_magic =
+        allocator.reserve_argspace(stride_magic.data(), stride_magic.size() * sizeof(int));
+    size_t idx_stride_shift =
+        allocator.reserve_argspace(stride_shift.data(), stride_shift.size() * sizeof(int));
+    size_t idx_reduced_strides =
+        allocator.reserve_argspace(reduced_strides.data(), reduced_strides.size() * sizeof(int));
+
+    // TODO: blending factors are not currently implemented
+    float alpha = 1.0f;
+    float beta = 0.0f;
+
+    int nthreads = static_cast<int>(shape_size(result_shape));
+
+    std::unique_ptr<gpu::primitive> softmax(new gpu::primitive{[=](void** inputs,
```

Comment:
It's a little bit confuse when build broadcast with softmax.

Reply:
Fixed.